library/src/main/java/com/devbrackets/android/exomedia/core/ListenerMux.kt

```diff
@@ -0,0 +1,460 @@
+
+    /**
+     * Retrieves if the player was prepared
+     *
+     * @return True if the player was prepared
+     */
+    var isPrepared = false
+        private set
+    private var notifiedCompleted = false
+    private var clearRequested = false
+
+    override fun onBufferingUpdate(mp: MediaPlayer, percent: Int) {
+        onBufferingUpdate(percent)
+    }
+
+    override fun onCompletion(mp: MediaPlayer) {
+        completionListener?.onCompletion()
+    }
+
+    override fun onError(mp: MediaPlayer, what: Int, extra: Int): Boolean {
+        return notifyErrorListener(NativeMediaPlaybackException(what, extra))
+    }
+
+    override fun onSeekComplete(mp: MediaPlayer) {
+        seekCompletionListener?.onSeekComplete()
+    }
+
+    override fun onPrepared(mp: MediaPlayer) {
+        notifyPreparedListener()
+    }
+
+    override fun onError(exoMediaPlayer: ExoMediaPlayer, e: Exception?) {
+        muxNotifier.onMediaPlaybackEnded()
+        muxNotifier.onExoPlayerError(exoMediaPlayer, e)
+        notifyErrorListener(e)
+    }
+
+    override fun onStateChanged(playWhenReady: Boolean, playbackState: Int) {
+        if (playbackState == Player.STATE_ENDED) {
+            muxNotifier.onMediaPlaybackEnded()
+
+            if (!notifiedCompleted) {
+                notifyCompletionListener()
+            }
+        } else if (playbackState == Player.STATE_READY && !isPrepared) {
+            notifyPreparedListener()
+        }
+
+        //Updates the previewImage
+        if (playbackState == Player.STATE_READY && playWhenReady) {
+            muxNotifier.onPreviewImageStateChanged(false)
+        }
+
+        //Clears the textureView when requested
+        if (playbackState == Player.STATE_IDLE && clearRequested) {
+            clearRequested = false
+            val clearableSurface = clearableSurfaceRef.get()
+
+            if (clearableSurface != null) {
+                clearableSurface.clearSurface()
+                clearableSurfaceRef = WeakReference<ClearableSurface>(null)
+            }
+        }
+    }
+
+    override fun onSeekComplete() {
+        muxNotifier.onSeekComplete()
+        seekCompletionListener?.onSeekComplete()
+    }
+
+    override fun onBufferingUpdate(@IntRange(from = 0, to = 100) percent: Int) {
+        muxNotifier.onBufferUpdated(percent)
+        bufferUpdateListener?.onBufferingUpdate(percent)
+    }
+
+    override fun onMetadata(metadata: Metadata) {
+        metadataListener?.onMetadata(metadata)
+    }
+
+    override fun onVideoSizeChanged(width: Int, height: Int, unAppliedRotationDegrees: Int, pixelWidthHeightRatio: Float) {
+        muxNotifier.onVideoSizeChanged(width, height, unAppliedRotationDegrees, pixelWidthHeightRatio)
+    }
+
+    // Analytics
+    override fun onPlayerStateChanged(eventTime: AnalyticsListener.EventTime?, playWhenReady: Boolean, playbackState: Int) {
+        analyticsListener?.onPlayerStateChanged(eventTime, playWhenReady, playbackState)
```

Comment:
With all these propagations, using delegation may help, with a noop fallback

Reply:
Absolutely

Reply:
Because the `analyticsListener` is variable doing a typical delegation won't work so I'm going to push this off until later so that we can move forward with other tasks

Reply:
Sure. I had a problem with this too. Mainly that it's mutable, since you can easily add it from a constructor with a default